src/tool/cppwinrt/strings/base_delegate.h

```diff
@@ -1,10 +1,61 @@
 
+WINRT_EXPORT namespace winrt
+{
+#if defined (WINRT_NO_MODULE_LOCK)
+
+    // Defining WINRT_NO_MODULE_LOCK is appropriate for apps (executables) that don't implement something like DllCanUnloadNow
+    // and can thus avoid the synchronization overhead imposed by the default module lock.
+
+    inline auto get_module_lock() noexcept
+    {
+        struct lock
+        {
+            constexpr uint32_t operator++() noexcept
+            {
+                return 1;
+            }
+
+            constexpr uint32_t operator--() noexcept
+            {
+                return 0;
+            }
+        };
+
+        return lock{};
+    }
+
+#elif defined (WINRT_CUSTOM_MODULE_LOCK)
+
+    // When WINRT_CUSTOM_MODULE_LOCK is defined, you must provide an implementation of winrt::get_module_lock()
+    // that returns an object that implements operator++ and operator--.
+
+#else
+
+    // This is the default implementation for use with DllCanUnloadNow.
+
+    inline std::atomic<uint32_t>& get_module_lock() noexcept
+    {
+        static std::atomic<uint32_t> s_lock;
```

Comment:
While we're here, let's improve the performance here by not forcing a full sequential consistency barrier on every ++ (which I believe is the default behavior for `std::atomic`.

Instead, let's apply the lessons we've learned and use a more performant "atomic_ref_count", with `std::memory_ordering::relaxed` ordering on `++`, and `std::memory_ordering::release` on `--`, with an added `acquire` thread fence in `--` when the ref count hits zero. I can resend an email I posted about this some time back if you want.

Reply:
Sure, that makes sense as we're already doing that for AddRef/Release.

Reply:
Done, I've added a `atomic_ref_count` helper as you suggested. Thanks!